modules/navier_stokes/src/executioners/SIMPLE.C

```diff
@@ -0,0 +1,1106 @@
+      &getUserObject<INSFVRhieChowInterpolatorSegregated>("rhie_chow_user_object"));
+  _rc_uo->linkMomentumSystem(_momentum_systems, _momentum_system_numbers, _pressure_tag_id);
+
+  // Initialize the face velocities in the RC object
+  _rc_uo->initFaceVelocities();
+
+  if (_pin_pressure)
+    _pressure_pin_dof = findDoFID("pressure", getParam<Point>("pressure_pin_point"));
+}
+
+void
+SIMPLE::relaxMatrix(SparseMatrix<Number> & matrix,
+                    const Real relaxation_parameter,
+                    NumericVector<Number> & diff_diagonal)
+{
+  // Zero the diagonal difference vector
+  diff_diagonal = 0;
+
+  // Get the diagonal of the matrix
+  matrix.get_diagonal(diff_diagonal);
+
+  // Create a copy of the diagonal for later use and cast it
+  std::unique_ptr<NumericVector<Number>> original_diagonal = diff_diagonal.clone();
+
+  // We cache the inverse of the relaxation parameter because doing divisions might
+  // be more expensive for every row
+  const Real inverse_relaxation = 1 / relaxation_parameter;
+
+  // Now we loop over the matrix row by row and sum the absolute values of the
+  // offdiagonal values, If these sums are larger than the diagonal entry,
+  // we switch the diagonal value with the sum. At the same time we increase the
+  // diagonal by dividing it with the relaxation parameter. So the new diagonal will be:
+  // D* = 1/lambda*max(|D|,sum(|Offdiag|))
+  // For more information see
+  //
+  // Juretic, Franjo. Error analysis in finite volume CFD. Diss.
+  // Imperial College London (University of London), 2005.
+  //
+  // The trickery comes with storing everything in the diff-diagonal vector
+  // to avoid the allocation and manipulation of a third vector
+  for (const auto row_i : make_range(matrix.row_start(), matrix.row_stop()))
+  {
+    std::vector<numeric_index_type> indices;
+    std::vector<Real> values;
+    matrix.get_row(row_i, indices, values);
+    Real abs_sum = std::accumulate(
+        values.cbegin(), values.cend(), 0.0, [](Real a, Real b) { return a + std::abs(b); });
+    Real abs_diagonal = std::abs(diff_diagonal(row_i));
+    Real new_diagonal = inverse_relaxation * std::max(abs_sum - abs_diagonal, abs_diagonal);
+    diff_diagonal.set(row_i, new_diagonal);
+  }
+
+  // Time to modify the diagonal of the matrix
+  for (const auto row_i : make_range(matrix.row_start(), matrix.row_stop()))
+    matrix.set(row_i, row_i, diff_diagonal(row_i));
+  matrix.close();
+
+  // Finally, we can create (D*-D) vector which is used for the relaxation of the
+  // right hand side later
+  diff_diagonal.add(-1.0, *original_diagonal);
+  diff_diagonal.close();
+}
+
+void
+SIMPLE::relaxRightHandSide(NumericVector<Number> & rhs,
+                           const NumericVector<Number> & solution,
+                           const NumericVector<Number> & diff_diagonal)
+{
+
+  // We need a working vector here to make sure we don't modify the
+  // (D*-D) vector
+  auto working_vector = diff_diagonal.clone();
```

Comment:
funny how you try hard not to allocate a third vector in the previous routine then just clone here :)

Reply:
I tried hard until I realized that most of the computational time goes towards the assembly instead of these little touches.